packages/react-dom/src/test-utils/ReactTestUtilsAct.js

```diff
@@ -61,18 +62,18 @@ function flushEffectsAndMicroTasks(onDone: (err: ?Error) => void) {
 
 function act(callback: () => Thenable) {
   let previousActingUpdatesScopeDepth;
+  let previousActingUpdatesSigil;
   if (__DEV__) {
     previousActingUpdatesScopeDepth = actingUpdatesScopeDepth;
+    previousActingUpdatesSigil = ReactShouldWarnActingUpdates.current;
     actingUpdatesScopeDepth++;
-    ReactShouldWarnActingUpdates.current = true;
+    ReactShouldWarnActingUpdates.current = ReactActingUpdatesSigil;
   }
 
   function onDone() {
     if (__DEV__) {
       actingUpdatesScopeDepth--;
-      if (actingUpdatesScopeDepth === 0) {
-        ReactShouldWarnActingUpdates.current = false;
-      }
+      ReactShouldWarnActingUpdates.current = previousActingUpdatesSigil;
```

Comment:
What happens if we throw in the middle? Does this get restored?

Reply:
yup, `onDone()` gets called for errors too, and passes the error upwards
https://github.com/facebook/react/blob/7caa1b2a0ef1d935971edeb202d776fbefd1ffa2/packages/react-dom/src/test-utils/ReactTestUtilsAct.js#L154-L155
https://github.com/facebook/react/blob/7caa1b2a0ef1d935971edeb202d776fbefd1ffa2/packages/react-dom/src/test-utils/ReactTestUtilsAct.js#L133-L134